sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala

```diff
@@ -64,6 +67,17 @@ class HDFSMetadataLog[T <: AnyRef : ClassTag](sparkSession: SparkSession, path:
     fileManager.mkdirs(metadataPath)
   }
 
+  protected val metadataCacheEnabled: Boolean
+  = sparkSession.sessionState.conf.getConf(SQLConf.STREAMING_METADATA_CACHE_ENABLED)
+
+  /**
+   * Cache the latest two batches. [[StreamExecution]] usually just accesses the latest two batches
+   * when committing offsets, this cache will save some file system operations.
+   */
+  protected[sql] val batchCache = Collections.synchronizedMap(new LinkedHashMap[Long, T](2) {
```

Comment:
Could we create `Map` instance only when `metadataCacheEnabled` is true? Maybe we can make it as `Option`

Reply:
I was going to say I wouldn't concern much about it since the instances of HDFSMetadataLog are not many. But another point would be that whether 1) we prefer having condition in if or 2) we prefer dealing with higher-level API e.g. foreach. Given that most usages are also having another condition for if statement, it doesn't seem to simplify the code.

Reply:
We could but the change is not really going to yield much difference.  The memory foot print of this is minimal.